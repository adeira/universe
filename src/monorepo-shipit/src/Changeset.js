// @flow strict

import { sprintf } from '@adeira/js';

type Diff = {
  +path: string,
  +body: string,
};

opaque type ChangesetData = {
  +id: string,
  +timestamp: string,
  +author: string,
  +subject: string,
  +description: string,
  +diffs: Set<Diff>,
  +coAuthorLines: $ReadOnlyArray<string>,
  +debugMessages: $ReadOnlyArray<string>,
};

export default class Changeset {
  declare id: string;
  declare timestamp: string;
  declare author: string;
  declare subject: string;
  declare description: string;
  declare diffs: Set<Diff>;
  coAuthorLines: $ReadOnlyArray<string> = [];
  debugMessages: $ReadOnlyArray<string> = [];

  isValid(): boolean {
    return this.diffs.size > 0;
  }

  getID(): string {
    return this.id;
  }

  withID(id: string): Changeset {
    return this.__clone({ id });
  }

  getTimestamp(): string {
    return this.timestamp;
  }

  withTimestamp(timestamp: string): Changeset {
    return this.__clone({ timestamp });
  }

  getAuthor(): string {
    return this.author;
  }

  withAuthor(author: string): Changeset {
    return this.__clone({ author });
  }

  getCoAuthorLines(): $ReadOnlyArray<string> {
    return this.coAuthorLines;
  }

  withCoAuthorLines(coAuthorLines: $ReadOnlyArray<string>): Changeset {
    return this.__clone({ coAuthorLines });
  }

  getSubject(): string {
    return this.subject;
  }

  withSubject(subject: string): Changeset {
    return this.__clone({ subject });
  }

  getDescription(): string {
    return this.description;
  }

  withDescription(description: string): Changeset {
    return this.__clone({ description });
  }

  getCommitMessage(): string {
    return `${this.getSubject()}\n\n${this.getDescription()}`;
  }

  getDiffs(): Set<Diff> {
    return this.diffs ?? new Set();
  }

  withDiffs(diffs: Set<Diff>): Changeset {
    return this.__clone({ diffs });
  }

  withDebugMessage(string: string, ...args: $ReadOnlyArray<string>): Changeset {
    const messages = this.debugMessages;
    return this.__clone({
      debugMessages: messages.concat(sprintf(string, ...args)),
    });
  }

  dumpDebugMessages(): void {
    console.log(sprintf('  DEBUG %s (%s)', this.getID(), this.getSubject()));
    for (const debugMessage of this.debugMessages) {
      console.log('    %s', debugMessage);
    }
  }

  __clone(newProps: { [$Keys<ChangesetData>]: $Values<ChangesetData>, ... }): Changeset {
    return Object.assign(Object.create(Object.getPrototypeOf(this)), this, newProps);
  }
}
