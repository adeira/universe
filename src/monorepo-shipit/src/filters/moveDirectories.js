// @flow strict

import Changeset from '../Changeset';

/**
 * Apply patches to a different directory in the destination repository.
 */
export default function moveDirectories(
  changeset: Changeset,
  mapping: Map<string, string>,
): Changeset {
  const rewriteCallback = oldPath => {
    let newPath = oldPath;
    for (const [src, dest] of mapping.entries()) {
      let matchFound = false;
      if (new RegExp(`^${src}`).test(newPath)) {
        matchFound = true;
      }
      newPath = newPath.replace(new RegExp(`^${src}`), dest);
      if (matchFound) {
        break; // only first match in the map
      }
    }
    return newPath;
  };

  const diffs = new Set();
  for (const diff of changeset.getDiffs()) {
    const oldPath = diff.path;
    const newPath = rewriteCallback(oldPath);
    if (oldPath === newPath) {
      diffs.add(diff);
      continue;
    }

    let body = diff.body;
    body = body.replace(new RegExp(`^--- a/${oldPath}`, 'm'), `--- a/${newPath}`);
    body = body.replace(new RegExp(`^\\+\\+\\+ b/${oldPath}`, 'm'), `+++ b/${newPath}`);

    diffs.add({
      path: newPath,
      body,
    });
  }

  return changeset.withDiffs(diffs);
}
