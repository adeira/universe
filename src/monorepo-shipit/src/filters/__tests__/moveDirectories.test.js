// @flow strict

import moveDirectories from '../moveDirectories';
import Changeset from '../../Changeset';

test.each([
  [
    'first takes precedence (first is more specific)',
    new Map([
      // from => to
      ['foo/public_tld/', ''],
      ['foo/', ''],
    ]),
    ['foo/orig_root_file', 'foo/public_tld/public_root_file'],
    ['orig_root_file', 'public_root_file'],
  ],
  [
    // this mapping doesn't make sense given the behavior, just using it to check that order matters
    'first takes precedence (second is more specific)',
    new Map([
      ['foo/', ''],
      ['foo/public_tld/', ''],
    ]),
    ['foo/orig_root_file', 'foo/public_tld/public_root_file'],
    ['orig_root_file', 'public_tld/public_root_file'],
  ],
  [
    'only one rule applied',
    new Map([
      ['foo/', ''],
      ['bar/', 'project_bar/'],
    ]),
    ['foo/bar/part of project foo', 'bar/part of project bar'],
    [
      'bar/part of project foo', // this shouldn't turn into 'project_bar/part of project foo'
      'project_bar/part of project bar',
    ],
  ],
  [
    'missing trailing slashes',
    new Map([
      ['foo', 'bar'],
      ['xyz/', 'aaa'],
    ]),
    ['foo/file', 'foo_baz/file', 'xyz/file'],
    [
      'bar/file',
      'bar_baz/file',
      'aaafile', // this can be a gotcha for someone
    ],
  ],
])('%s', (testName, mapping, inputPaths, expected) => {
  const changeset = new Changeset().withDiffs(
    new Set(inputPaths.map(path => ({ path, body: 'placeholder' }))),
  );
  const diffs = moveDirectories(changeset, mapping).getDiffs();
  expect([...diffs].map(diff => diff.path)).toEqual(expected);
});
