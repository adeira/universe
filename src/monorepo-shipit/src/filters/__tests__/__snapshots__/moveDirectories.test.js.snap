// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles complex filenames requiring regexp escaping correctly 1`] = `
Changeset {
  "coAuthorLines": Array [],
  "debugMessages": Array [],
  "description": "MOCKED_HEADER",
  "diffs": Set {
    Object {
      "body": "new file mode 100644
index 0000000000000000000000000000000000000000..fb7887c2e2ec937f47e9a58a22ffc8fd705379ff
--- /dev/null
+++ b/react-native/ios/YaComiste/Images.xcassets/Logo (no background).imageset/Contents.json
@@ -0,0 +1,21 @@
+{
+  \\"images\\" : [
+    {
+      \\"idiom\\" : \\"universal\\",
+      \\"scale\\" : \\"1x\\"
+    },
+    {
+      \\"idiom\\" : \\"universal\\",
+      \\"scale\\" : \\"2x\\"
+    },
+    {
+      \\"filename\\" : \\"100.png\\",
+      \\"idiom\\" : \\"universal\\",
+      \\"scale\\" : \\"3x\\"
+    }
+  ],
+  \\"info\\" : {
+    \\"author\\" : \\"xcode\\",
+    \\"version\\" : 1
+  }
+}
",
      "path": "react-native/ios/YaComiste/Images.xcassets/Logo (no background).imageset/Contents.json",
    },
  },
}
`;
